### ISSUE TITLE
Speed-up bonds detection with an octree

### ISSUE THREAD

**Issue by Luthaf** [chemfiles/chemfiles#116]
The idea comes from this github project: https://github.com/PorridgeBear/imvs/blob/master/IMVS/AtomCloud.swift.

Creating an octree will allow for faster lookup in bonds detection, replacing the brute-force 2D loop here: https://github.com/chemfiles/chemfiles/blob/3cbd6d78bc99ded786044eee6db7022ed2bf4c75/src/Frame.cpp#L60-L80

Instead of looping over all the atoms in the frame, we will only need to loop over the atoms in the surrounding cells from the tree.

**Comment by richardjgowers** [comment]
@Luthaf fwiw, we had a GSOC student last year looking at domain decomposition methods for limited distance searches (ie distances up to a threshold, here bond lengths). The full details are [in his blogposts](https://ayushsuhane.github.io/), but ultimately we found using a rip off of Gromacs' neighbour list algorithm was faster than a kd-tree/octree.  We were using python bindings of these methods, so maybe your mileage may vary!

**Comment by Luthaf** [comment]
Yeah, cells lists (that the blog post mention rather than neighbours lists) might be a good idea here. Thanks for the head up!
